### ISSUE TITLE
Add `strict` mechanism for opting into stricter subsets of the language 

### ISSUE THREAD

**Issue by Keno** [JuliaLang/julia#54903]
We've had a few discussions the past few weeks about a feature tentatively dubbed `pragma strict` after similar constructs in other languages. However, there wasn't really a cohesive writeup of the intent, so triage asked me to write one up to serve as the basis for discussion and fleshing out. I intend to edit this issue as the idea evolves.

## Basic idea

The basic idea of the `pragma strict` feature is to have an opt-in mechanism of turning julia programs that are semantically valid, but undesirable for other reasons (e.g. using ambiguous syntax that should have arguably been disallowed, but we can't for backwards compatibility reasons) into errors. This would be an opt-in feature for developers who have personal, organizational or regulatory requirements for requiring stricter coding standards. An additional motivation is to provide an additional vehicle for low-frictition language evolution. For example, if a specific opt-in turns out to be popular across the majority of packages, a potential julia 2.0 that made the opt-in automatic while technically breaking, would be largely non-breaking in practice.

We are not imagining a single `strict mode` opt in here, but rather a finer grained set of options, plus versioned collections of options for particular use cases. See the last section for a an initial list of such options.

It is worth emphasizing again that this feature is only intended to disallow undesirable programs that are otherwise semantically valid. It is not intended to cause meaningful semantic differences in programs that are valid both in standard semantics and under the opt-in restrictions (i.e. turning on the restrictions may cause things to error, but if they don't the program should behave the same).

## How does the opt-in work?

One of the primary questions in this proposal is how the user expresses the opt-in. There's a few separate semantic options, each
with a number of potential syntax options.

1. Per module opt-in like our existing `Experimental.@compiler_options`
2. Per file opt-in (e.g. using a magic comment on the first line) - popular in some other languages
3. Per project opt-in in Project.toml

After some discussion on triage, a Project.toml-level opt-in seems like the best option. The primary motivation here is to allow opt-ins that need to be done in the parser (e.g. whitespace requirements). We don't currently define the execution ordering of parsing and execution for packages, so a module-toplevel opt-in may be semantically too late (relatedly, it may be ambiguous what happens when the opt-in is placed in the middle of a disallowed parse). An additional concern is that ideally IDE tooling would be able to understand the active set of restrictions without having to look at the code.

## Concrete Project.toml syntax options

One convenient option would be reusing Preferences.jl. One might imagine a julia-level preference like:

```
name = "MyPackage"

[preferences.julia]
strict = ["nomultiassign", "uniqueidentifiers"]
```

This doesn't fully mesh with the usual preferences semantics, since preferences are ordinarily uniqued per-UUID while
they would be private for a particular package, but this might be ok. Alternatively, we could reserve the `strict` key
in each individual package's preference table:

```
name = "MyPackage"

[preferences.MyPackage]
strict = ["nomultiassign", "nolocalshadow", "noglobalshadow"]
```

Alternatively, we could have a new top-level `strict` section:
```
name = "MyPackage"
[strict]
julia = ["nomultiassign", "nolocalshadow", "noglobalshadow"]
```
## Initial idea list for opt-in options

In this section, I'm collecting a list of potential options that might be implemented. However, I am not at this point asking people to brainstorm all the possibilities that could be implemented. I'm also not asking for detailed discussion on what should or should not be included in a particular option. Rather, I wanted to have a place to list all the ideas that have already come
up and a place to link any issues that could be addressed by this feature. Full design discussions for individual flags can be had on the PRs to implement them once the overall mechanism is in place.

### Individual options

- `nomultiassign`
	
Disallows multiple assignments in the same expression without parantheses. I.e. disallows `a = b, c = d, e, = f = (1, 2)`

- `nolocalshadow`

Disallows shadowing of local variables, e.g. in the following

```

function foo()

	for i = 1:10

		for i = 1:10 # Error shadowing local `i` 

		end

		all(1:10) do i # Error shadowing local `i`
			iszero(i)
		end
	end
end
```

- `noglobalshadow`

Disallows shadowing of global variables, e.g. in the following:
```
function foo()
	missing = false # Error local `missing` shadows imported global `missing`
end
```

- Some variant of unique assignment

Stefan had proposed introducing a unique assignment operator, e.g. `:=` for which there would then be a corresponding opt-in to enforce all assignments use it

- Enforce export versioning

If we implement some variant of export versioning, there could be an opt-in forbidding unversioned exports.

### Collections

The idea of collections is that users in general don't want to individually decide which opt ins matter to them, but will likely be following a standard set by their organizations or prescribed by a style guide. To this end, there could be meta opt-ins like "basestyle", which would
activate a standard collection of opt-ins. These collections should be versioned and activated based on the min-compat version of Julia. In this way, new opt-ins can be added to a collection, without automatically activating them on a julia version upgrade.

**Comment by adienes** [comment]
I would suggest that one of the individual options should be disallowing control flow in non-"statement" position, i.e. https://github.com/JuliaLang/julia/issues/50415#issuecomment-2053547922

**Comment by jariji** [comment]
https://github.com/JuliaLang/julia/issues/51223 is my proposal for `:=` reassignment.

**Comment by jariji** [comment]
I like (Stefan's?) idea that whitespace must match operator precedence so you can't write `2 * 3+1`, you have to write `2*3+1` or `2*3 + 1` or `2 * 3 + 1`.

**Comment by StefanKarpinski** [comment]
Another thing we had talked about was having a set of defaults based on a version, which you could add to or subtract from which might look like this:
```toml
strict = ["1.12 defaults", "no local shadow", "-explicit imports"]
```

**Comment by nsajko** [comment]
Some of the configurations should probably disallow accessing non-public names, including:

1. names of instance properties
2. names in a module

IMO this should be opt-out for all packages, but not for the REPL.

**Comment by jariji** [comment]
Just an idea - What do you think about having these options be installable packages so users can contribute their own rules, rather than having an official set of rules?

**Comment by davidanthoff** [comment]
One question is whether this needs to be implemented in Julia itself, or whether this belongs more in a linter like tool. At some level it strikes me that if this is a thing, we most definitely would want to implement support for this in things like the language server. And then the question: what is gained by having two implementations?

**Comment by nsajko** [comment]
@davidanthoff maybe I'm wrong, but I have a feeling you may have this backwards. My feeling is that the only way for the language server to become a clear win for users (currently it's quite annoying with the false positive warnings) is to plug into the Julia implementation quite directly, maybe similarly to Cthulhu.jl. So maybe the language server for Julia should be just a thin wrapper around Julia.

**Comment by davidanthoff** [comment]
@nsajko probably best to stick with Keno's suggestion to collect ideas here but not discuss or evaluate them in detail, that would presumably just distract from the topic of this issue. Having said that, if you have ideas and thoughts about the LS, please open an issue over at it's repo and we can discuss there.

**Comment by LilithHafner** [comment]
I agree that per-project is the best approach.

I agree that Project.toml is the place to put this opt-in and configuration.

## Concrete Project.toml syntax feedback

I think a toplevel `strict` is the simplest approach to avoid confusion over subtle inconsistencies with Preferences.jl's preference resolution.

```toml
name = "MyPackage"
strict = ["noreassignment", "nomisleadingwhitespace"]
```

Additionally, this strictness is a property tied to the package about as closely as it's name and version—it's likely that a project declared without strict rules will fail to parse with them. The closest analog to this feature I know of is [Rust editions](https://doc.rust-lang.org/cargo/reference/manifest.html#the-edition-field). In Rust, that field is stored in the `[package]` table, which serves a analogous role as the toplevel table in our Project.toml files.

Reserving the strict key in each individual package's preferences is a bit breaking. It's also unclear what it means to set the strict preference of any package other than the one named by the Project.toml file. Syntax that enables this seems problematic:

```toml
name = "MyPackage"

[preferences.OtherPackage]
strict = ["nomultiassign", "nolocalshadow", "noglobalshadow"]
```

A toplevel `[strict]` section seems unnecessarily verbose compared to a toplevel `strict` key.

**Comment by c42f** [comment]
> provide an additional vehicle for low-frictition language evolution

Yes! We need this for syntax evolution which is often technically breaking but not actually very breaking at all in practice. There's so many examples of this. Some being #36547 and #54915.  In https://github.com/JuliaLang/julia/issues/36547#issuecomment-1449143117 I show that several bugs would be fixed by this syntax change. But the change itself is nevertheless, technically breaking and it's a really tough call to decide whether to do it.

Lilith has already mentioned Rust Editions. A core part of Rust editions are that they don't bifurcate the ecosystem because crates with using different editions can work together. We should definitely do that to avoid a python 2/3 style debacle.

> For example, if a specific opt-in turns out to be popular across the majority of packages, a potential julia 2.0 that made the opt-in automatic while technically breaking, would be largely non-breaking in practice

For a lot of minor syntax improvements, I think they'd best be expressed as "use the latest syntax as of Julia version 1.x" rather than as opt-in flags. If we're trying to improve the syntax to change/remove ambiguous or confusing syntactic constructs, we want an incentive for the whole ecosystem to drop the old syntax. For example if we do the change @JeffBezanson mentioned here https://github.com/JuliaLang/julia/issues/54915#issuecomment-2235198794 in a Julia edition we do want packages to drop the old syntax as soon as possible.

"Use Julia 1.x syntax edition" is great from this point of view:
- Users have an *incentive* to opt in because they get the latest syntax niceties (for example multidimensional array literals)
- They pay the minor price of updating old syntax (can be automated!)
- Users of the ecosystem benefit from most packages being on the latest syntax version

So I think "syntax evolution" should not be fine grained, where at all possible - it's a bit different from the other "strict mode" things which are mentioned above, where users might want fine-grained control over opting out of certain language constructs.

**Comment by nsajko** [comment]
xref #43654 #46411 #52014 #55304

**Comment by nsajko** [comment]
xref #50040

**Comment by c42f** [comment]
https://github.com/JuliaLang/julia/issues/48434 seems like another candidate for Julia Editions - not so much "strict mode", but a change to lowering semantics. It's a reasonable candidate because we can detect most semantic changes introduced by changing the `try` scoping rules and provide help for people porting their code either as a diagnostic or automated refactoring pass. (The tricky part would be changes to semantics in top-level code - we can probably still deal with this using static analysis for most packages but there will always be hard cases.)

**Comment by o314** [comment]
rustc compiler handles things this way. see
https://rustc-dev-guide.rust-lang.org/implementing_new_features.html
https://rustc-dev-guide.rust-lang.org/feature-gates.html
interop with the ast is handled too
https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast_passes/feature_gate/fn.check_crate.html
there is macro for this (w rust macro_rules eg composable matchable macro )

**Comment by MilesCranmer** [comment]
What about using a scoped approach like `@stable`/`@unstable` in [DispatchDoctor.jl](https://github.com/MilesCranmer/DispatchDoctor.jl)? Like:

```julia
@strict :nomultiassign :nolocalshadow begin

#= package code =#

end
```

It recurses through `include` and submodules too.

This way you could update parts of your code at once. And perhaps sometimes you just need to disable it just for one single function, so you can turn it off with

```julia
@nostrict begin

end
```

which disables the outermost `@strict` for the scope.

You can use Preferences.jl to pass defaults for the `@strict` scopes so you can configure it from the top level. This is the same way DispatchDoctor.jl uses Preferences.jl to provide defaults for codegen level and how many `Union`s to consider types unstable, which has worked really well I think.

**Comment by younes-io** [comment]
Another way to reconcile strict modes, linter rules, and "edition-like" changes is to introduce a graduated enforcement system: a "ladder" of increasingly strict enforcement. This would let developers climb as high up the ladder as they can, without forcing them to jump all at once. For instance:

**Lint-Only (Soft)** : Violations would merely produce warnings or style hints in IDEs and CI, and it would be useful for early detection and adopting best practices without breaking code.

**Strict-Warn (Medium)** : the compiler (or a compiler-like pass) can still compile the code but issues deprecation-level warnings that are treated seriously in automated checks. This could also be great for codebases that want warnings to become "CI failures", but not show-stoppers in local dev if you’re prototyping something quickly..

**Strict-Error (Strong)**: at this level, strictness violations become hard errors: the code simply does not run under your chosen strict/edition setting. Could work well for teams or organizations that want compliance guarantees (e.g. regulatory), or for new library code that aims to adopt "modern Julia".

Crucially, one can imagine each rung on the ladder being toggled in `Project.toml`. So if you specify `strict = ["julia-2025:strict-warn"]`, you get the medium rung for all the 2025 changes. This keeps the ecosystem from fragmenting.. people can choose their rung depending on how ready they are to enforce those rules...

Obviously, this extends the notion of a "Julia edition" to let project owners define not just which strict or edition rules to adopt, but also how strongly to enforce them, all in one place..

**Comment by nsajko** [comment]
xref #57311

**Comment by LEXUGE** [comment]
My two cents on what should probably go inside a strict mode:

I think one should avoid characterize the strict mode by putting together many small admonishments (e.g. no argument should have type `Any`). That's what linter does and it would make things unclear about what strict mode promises.

Instead, I think it's better to structure strict mode around some "cohesive" promise (e.g. make code more "inferable" by requiring type annotation, not allow dynamic dispatch except on those labelled abstract types, etc). I hope working out what to add to strict mode in this way will give a more concrete sets of rules to enforce and relatively few "strict mode flags" to turn on, instead of a long list of admonishments.

**Comment by FabrizioTorrico** [comment]
Waiting for this 🙏

**Comment by el-oso** [comment]
What about a `trimonly` option that disallows anything that potentially doesn't work with `juliac --trim`

**Comment by rsnumerics** [comment]
Can I suggest to look into how SPARK was created as a strict subset of Ada? As a mathematician I like the idea that code correctness can be automatically proven for SPARK against a formal specification. Some industry applications (aerospace, healthcare) demand demonstrable code correctness.

**Comment by c42f** [comment]
Just to cross link a related PR - some initial support for syntax evolution is now available as of https://github.com/JuliaLang/julia/pull/60018. We probably need to refine how macro interop will work to avoiding splitting the ecosystem but there's a good mechanism for parts of it already.

(This is different from strict mode - see #60180 for that)

**Comment by bbrehm** [comment]
Some more thoughts:

1. "Surface level": I think we should be clear on what level strictness applies. I think we should apply on surface level (i.e. the actual text content), on expression/AST level, and on lowered linear-IR level. We should _not_ apply on anything later, i.e. on things that can get invalidated. The difference between surface-level and AST-level is that `for i=1:10 end` and `for i in 1:10` are the same as AST and different on surface.
2. Generated functions. There should be some mechanism to opt into and out of strictness-checking for the generated expression. A strictness failure in a generated expression/function naturally cannot blow up before runtime / once type arguments are known. So the bug/failure might be in the library that opts into strictness, but the error will appear for the user of the library.
3. Stability. We need to be careful with strict-mode options. Future julia versions can make old strict options more permissive, but cannot make them even more strict. So we should only make promises that we can keep.
4. `@eval module code` and similar: We need to be clear on whether `module`'s strictness settings apply.

Some more interesting strictness options:

- `no_capture_box`: Disallows closure capturing via `Core.Box`. This is something a lot of people would like. But it is problematic with respect to stability: We want (1) code that successfully lowers under `no_capture_box` will never ever closure-capture via a box. (2) Code that successfully lowers under `no_capture_box` in 1.x must do so in 1.(x+n). So simply checking the lowered IR for capture-boxes is too relaxed: It prevents us from any future changes in lowering that introduce more boxes. Instead, we need to come up with some set of sufficient conditions that prevent closure-capture-boxes, and commit to always elide captures under these conditions. (we should still check whether the generated IR has closure-capture-boxes; if we find any, we need to display an error message that points to https://github.com/JuliaLang/julia/issues/new).

- `no_implicit_convert`: Implicit `convert` is a source for issues that are surprising for newcomers. Strictness cannot change semantics, apart from introducing strictness lowering errors. Hence, this effectively opts out of type declarations on variables / functions a la `f()::Int = 1.0` that are easy to mistake for type-asserts. This leaves ccall / cconvert / unsafe_convert as-is -- I see no reasonable way of making that more explicit in a fully backwards compatible way.

- `restrict_goto`: Further restrict `@goto`. We already restrict goto around try/catch/finally. Useful possible further restrictions are: 1. ensure reducible CFG, i.e. never jump into loops (easy to check + maintain on lowered IR), and 2. never jump into the middle of scopes (easy to check on AST). The two are related -- cf webassembly, java, keyword "semi-structured". Maybe we want both options. I tend to find that labeled break / continue in java are powerful enough most of the time and not too inconvenient, but mere unlabeled break/continue are not expressive enough. `@goto` is very convenient; it would be nice to have some optional guard-rails that help verify that most mundane uses are really just syntactic sugar over the labeled-continue/break offered in other languages. Also possibly important for web-assembly. I'm not sure how how long in the compilation pipeline reducibility of the CFG is maintained (during julia SSA-IR optimizations? during llvm passes?), nor how much it helps the compiler. But I do know that I don't want to write irreducible CFG by oversight, only with considered intention.
 
- `no_undef_locals`: Forbid undefined local variables. This option is similiar in nature to `no_capture_box`: It should guarantee that no UndefVarError "in local scope" can be raised and that no runtime check for defined-ness can be emitted. Checking the lowered code is currently unhelpful: Knowledge about this is only available once we are in SSA-form / have removed slots. The obvious check would be: Do CFG paths exist along which the variable is unassigned?

`no_undef_locals` is especially relevant for try/catch -- it is too easy to write catch/finally handlers that explode because they access variables that were supposed to be initialized inside the try block. The other try/catch potential issue is with respect to scope... but that is easier to discover in testing. Whereas `try x = f(); g!(x); catch ex; logerror(ex, x); recover_something() end` is easy to write and test to the point that it deals with exceptions thrown from `g!` (i.e. x exists in the outer scope); and then the error logging throws UndefVarError from inside the exception handler, because surprisingly `f()` threw.

**Comment by c42f** [comment]
> no_capture_box: Disallows closure capturing via Core.Box.

I'd like an easy way to do this check, in some form. It's way too easy to think you've written code with no boxes only to find you slipped up somewhere.

> It prevents us from any future changes in lowering that introduce more boxes

I don't think we'd consider language changes which force us to use more boxes. I can't actually imagine what those might be; the `Box`s are consequences of lexical scope and those scope rules have been fixed pretty much forever.

> `no_undef_var` ... Checking the lowered code is currently unhelpful

Yes. Lowering does emit `NewvarNode` which can be a partial hint but lowering's analysis of control flow is too weak for this to be very useful. Globals can still throw `UndefVarError` so maybe this would be `no_undef_locals`.

**Comment by c42f** [comment]
Note that a rough future for how strict mode might be implemented is recorded at https://github.com/JuliaLang/julia/pull/60180#issuecomment-3560537336

**Comment by bbrehm** [comment]
>I don't think we'd consider language changes which force us to use more boxes. I can't actually imagine what those might be; the Boxs are consequences of lexical scope and those scope rules have been fixed pretty much forever.

It's not about language changes that force us to use more boxes.

The issue is: The language implementation (flisp lowering, julialowering) uses a lot of boxes that are not forced by the language. They use a bunch of more or less naive heuristics to avoid some boxes. Fun effects if these implementations diverge, cf eg https://github.com/JuliaLang/julia/pull/60567

I could make a very simple PR to remove the boxing in
```
julia> function foo()
       x=0
       @label baz
       fun = ()->x
       return fun()
       end
```
Namely: Don't bail out of the capture box optimization heuristics due to `@label` if there is no `@goto`. Then, you come along and say "this is stupid, let's remove that optimization -- no real code has unused `@label`". And then, with strict-mode `no_capture_box`, this would be a breaking change.

There must be some room for the compiler/implementation to make tradeoffs --some gap between things that the compiler is obligated to do and things that the compiler is allowed to do.

With respect to capture box optimization, there are currently no obligations. I would like that to change. But closing this gap and going "Everything that is not forbidden is mandatory" would be insane.

PS.
>so maybe this would be no_undef_locals

That's what I meant. Thanks, this is a much better name than `no_undef_var`! I edited my comment.
